includes/distributor-customizations/class-authorship-filters.php

```diff
@@ -0,0 +1,104 @@
+<?php
+/**
+ * Newspack Network authorship filters.
+ *
+ * @package Newspack
+ */
+
+namespace Newspack_Network\Distributor_Customizations;
+
+use Distributor\DistributorPost;
+use Newspack\Data_Events;
+use Newspack_Network\Debugger;
+use Newspack_Network\User_Update_Watcher;
+use Newspack_Network\Utils\Users as User_Utils;
+
+/**
+ * Class to handle authorshipt filters
+ *
+ * This class is responsible for filtering the authorship data that is publicly displayed in the site using the distributed authors and including CAP's guest authors
+ */
+class Authorship_Filters {
+
+	/**
+	 * Initializes the class
+	 *
+	 * @return void
+	 */
+	public static function init() {
+		add_filter( 'get_coauthors', [ __CLASS__, 'filter_coauthors' ], 10, 2 );
+		add_filter( 'newspack_author_bio_name', [ __CLASS__, 'newspack_author_bio_name' ], 10, 3 );
+	}
+
+	/**
+	 * Filters the coauthors of a post to include the distributed authors and CAP's guest authors
+	 *
+	 * @param array $coauthors Array of coauthors.
+	 * @param int   $post_id   Post ID.
+	 * @return array
+	 */
+	public static function filter_coauthors( $coauthors, $post_id ) {
+		if ( ! class_exists( 'Distributor\DistributorPost' ) ) {
+			return $coauthors;
+		}
+
+		// We don't want to filter authors on admin, as it might break things.
+		if ( is_admin() ) {
+			return $coauthors;
+		}
+
+		// Only filter posts that are still linked to the original post.
+		$distributor_post = new DistributorPost( $post_id );
+		if ( ! $distributor_post->is_linked ) {
+			return $coauthors;
+		}
+
+		$distributed_authors = get_post_meta( $post_id, 'newspack_network_authors', true );
+
+		// If anything, at any point, goes wrong with the distributed authors, we return the original coauthors.
+		if ( ! $distributed_authors ) {
+			return $coauthors;
+		}
+
+		$filtered_coauthors = [];
+
+		foreach ( $distributed_authors as $distributed_author ) {
+			if ( ! isset( $distributed_author['type'] ) ) {
+				return $coauthors;
+			}
+
+			if ( 'guest_author' === $distributed_author['type'] ) {
+				$filtered_coauthors[] = (object) $distributed_author;
```

Comment:
I don't think this produces the schema expected from the `get_coauthors()` method. I'm getting a lot of warnings from functions that expect a `user_login`, `user_nicename` property and `ID` (in caps) instead of `id`:

I'm aware I'm not running https://github.com/Automattic/newspack-theme/pull/2215 in this test but I don't think we should break what is returned from the function. This can lead to a lot of unexpected things breaking.

```
PHP Warning:  Undefined property: stdClass::$user_login in /Users/miguelpeixe/devel/newspack-ads/includes/providers/gam/class-gam-model.php on line 1128
PHP Warning:  Undefined property: stdClass::$ID in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 29
PHP Warning:  Undefined property: stdClass::$user_nicename in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 29
PHP Warning:  Undefined property: stdClass::$ID in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 38
PHP Warning:  Undefined property: stdClass::$user_nicename in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 38
PHP Warning:  Undefined property: stdClass::$ID in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 39
PHP Warning:  Undefined property: stdClass::$ID in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 61
PHP Warning:  Undefined property: stdClass::$user_nicename in newspack-theme/newspack-theme/template-parts/post/author-bio.php on line 61
```

For reference, this is what this line is producing for my guest author:
```
stdClass Object
(
    [type] => guest_author
    [id] => 158
    [display_name] => John Doe
    [user_email] => john@doe.com
    [website] => https://example.com
    [first_name] => John
    [last_name] => Doe
    [description] => John is a guest author
    [newspack_job_title] =>
    [newspack_role] =>
    [newspack_employer] =>
    [newspack_phone_number] =>
)
```

Reply:
> I'm aware I'm not running https://github.com/Automattic/newspack-theme/pull/2215 in this test but I don't think we should break what is returned from the function. This can lead to a lot of unexpected things breaking.

Yes, good call, I'll replicate the object with the same properties - and some additional properties shouldn't be a problem

Reply:
Fixed in https://github.com/Automattic/newspack-network/pull/25/commits/11a3fa3eec0b2563895e2978d532e22fb3b1f469

Notice we still need to remove the contents of `user_nicename` so we know we don't want to render a link for this author. But all the properties are there.

You'll need to redistribute a post with a guest author for the changes to make effect in the new distributed post. You can use the same existent guest author though

Reply:
This is the only thing this PR doesn't resolve on its own. The theme will render a broken link for the guest author. Always.

Reply:
One idea, not very elegant, but we could:

* modify the ID of the users to something easily identifieable like `-2` or `999998989`. (or do this with the user_nicename)
* Add a filter to `author_link` and return `#` for these authors
* At least the link would not point to anywhere wrong...

WDYT?

Reply:
This is how it looks like -> https://github.com/Automattic/newspack-network/pull/30

Reply:
Moreover: I think it's a good idea to set an invalid ID for guest authors, otherwise it could easily conflict with an existing author in the site

Reply:
Why not `-1`? It's a common placeholder for "not found".

Reply:
I wanted to be unique, not to get confused with anything else

Reply:
What else could `-1` be?

Reply:
> What else could -1 be?
Not found :) 

It's not "not found", it's "filtered by Newspack Network"

Reply:
It's "not found" on that site as an entity, but `-2` is fine.

Reply:
But what do you think about the replacement for `#`?

I think it's terrible, but at least makes this feature more self-contained in this plugin. The theme can improve the support for this feature, but it will at least work better with any theme

Reply:
I think `#` is a good solution. Is linking back to the guest author page on the original site a bad idea?

Reply:
That's another option. But we discussed this on that call that day and decided against it.

We didn't want to encourage its use to get this behavior "If you want to link back to the original site use a guest author, and if you want to link to the target site, use a regular user"

Reply:
Good point! `#` is solid, the theme can `$author_url !== '#'` to condition the rendering of the `<a />` and others will just have a somewhat harmless empty hash link...

Reply:
Anything else here? :)